connect/src/com/telenor/mobileconnect/MobileConnectSdkProfile.java

```diff
@@ -112,23 +115,74 @@ public Uri getAuthorizeUri(Map<String, String> parameters, List<String> locales)
+                            return;
+                        }
+                        callback.onError();
+                    }
 
-        if (odResult != null) {
-            parameters.put("login_hint", String.format("ENCR_MSISDN:%s", odResult.getSubscriberId()));
-        }
+                    @Override
+                    public void failure(RetrofitError error) {
+                        callback.onError();
+                    }
+                };
 
-        if (isInitialized || initialize(odResult)) {
-            return DoNext.proceed;
-        }
+        final Callback<OperatorDiscoveryAPI.OperatorDiscoveryResult> odCallbackForMsisdn =
+                new Callback<OperatorDiscoveryAPI.OperatorDiscoveryResult>() {
+                    @Override
+                    public void success(OperatorDiscoveryAPI.OperatorDiscoveryResult operatorDiscoveryResult, Response response) {
+                        parameters.put("login_hint", String.format("ENCR_MSISDN:%s", operatorDiscoveryResult.getSubscriberId()));
+                        odCallbackForMccMnc.success(operatorDiscoveryResult, response);
+                    }
+
+                    @Override
+                    public void failure(RetrofitError error) {
+                        if (isInitialized) {
+                            callback.onSuccess();
+                            return;
+                        }
+
+                        TelephonyManager phMgr = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
+                        String networkOperator = phMgr.getNetworkOperator();
 
-        return DoNext.cancel;
+                        if (TextUtils.isEmpty(networkOperator)) {
+                            callback.onError();
+                            return;
+                        }
+
+                        if (!TextUtils.isEmpty(networkOperator)) {
+                        final String mcc = networkOperator.substring(0, 3);
+                        final String mnc = networkOperator.substring(3);
+                        getOperatorDiscoveryApi().getOperatorDiscoveryResult_ForMccMnc(
+                                getOperatorDiscoveryAuthHeader(),
+                                operatorDiscoveryConfig.getOperatorDiscoveryRedirectUri(),
+                                mcc,
+                                mnc,
+                                odCallbackForMccMnc);
+                        }
+                    }
+                };
+
+        String msisdn = readPhoneNumber();
+        if (msisdn != null) {
+            getOperatorDiscoveryApi().getOperatorDiscoveryResult_ForMsisdn(
+                    getOperatorDiscoveryAuthHeader(),
+                    new OperatorDiscoveryAPI.BodyForMsisdn(
+                            operatorDiscoveryConfig.getOperatorDiscoveryRedirectUri(),
+                            msisdn),
+                    odCallbackForMsisdn);
+        } else {
+            odCallbackForMsisdn.failure(null);
```

Comment:
Yes, but I guess you could refactor the logic that happens in failure and put it in a mccMncFlow() method for example and then call this both from the onFailure and from the else branch.
Calling .failure to do your happy path is really hard to parse, for me at least.